lib/CL/clCreateSubDevices.c

```diff
@@ -38,37 +40,117 @@ POname(clCreateSubDevices)(cl_device_id in_device,
                            cl_device_id *out_devices,
                            cl_uint *num_devices_ret) CL_API_SUFFIX__VERSION_1_2
 {
-   int errcode;
-   cl_device_id sub1 = in_device;
-   cl_device_id sub2 = in_device;
+   cl_int errcode = CL_SUCCESS;
+   cl_uint count_devices = 0;
+   cl_device_id *new_devs = NULL;
+   // number of elements in (copies of) properties, including terminating null
+   cl_uint num_props = 0;
+   uint i;
 
    POCL_GOTO_ERROR_COND((in_device == NULL), CL_INVALID_DEVICE);
+   POCL_GOTO_ERROR_COND((properties == NULL), CL_INVALID_VALUE);
+   POCL_GOTO_ERROR_COND((num_devices && !out_devices), CL_INVALID_VALUE);
+   POCL_GOTO_ERROR_COND((!num_devices && out_devices), CL_INVALID_VALUE);
+
+   /* check that the partition property is supported by the device */
+   errcode = CL_INVALID_VALUE;
+   for (i = 0; i < in_device->num_partition_properties; ++i) {
+     if (properties[0] == in_device->partition_properties[i]) {
+       errcode = CL_SUCCESS;
+       break;
+     }
+   }
+   if (errcode != CL_SUCCESS)
+     goto ERROR;
+
+   /* Ok, it's a supported partition property, count the number of devices; currently,
+    * we only support EQUALLY and BY_COUNTS, which enumerate the number of devices
+    * differently */
+   if (properties[0] == CL_DEVICE_PARTITION_EQUALLY) {
+     // error out if the number of CUs per device is 0 or bigger than the number of
+     // CUs of in_device
+     POCL_GOTO_ERROR_COND(
+       (properties[1] == 0 || properties[1] > in_device->max_compute_units),
+       CL_INVALID_VALUE);
+     // error out if properties isn't zero-terminated
+     POCL_GOTO_ERROR_COND(properties[2] != 0, CL_INVALID_VALUE);
+
+     count_devices = in_device->max_compute_units / properties[1];
+     num_props = 3; // partition type, CUs per device, terminating 0
+   } else {
```

Comment:
Shouldn't this be else if ( ... Cl_DEVICE_PARTITION_BY_COUNTS)  and the actual else to return 0 due to an unsupported partitioning type?